### PR TITLE
Mr jonatans/issue 6

### PR DESCRIPTION
**Pull request by MrJonatans**

### CODE REVIEW

File: src/kaitai_spec/attribute/enumeration.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            match endian_unchecked {
+                "le" => endian = Endian::Little,
+                "be" => endian = Endian::Big,
+                "" => {}
+                _ => return Err(A::Error::custom("invalid type")),
+            }
+        }
+        2 => {}
+        _ => return Err(A::Error::custom("invalid type")),
+    }
```
Comment (ArkadiyParovozow) [inline src/kaitai_spec/attribute/enumeration.rs:65]:
```suggestion
    let endian = match type_unchecked.len() {
        4 => {
            match &type_unchecked[2..4] {
                "le" => Some(Endian::Little),
                "be" => Some(Endian::Big),
                _ => return Err(A::Error::custom("invalid type")),
            }
        }
        2 => None,
        _ => return Err(A::Error::custom("invalid type")),
    };
```

File: src/kaitai_spec/attribute/enumeration.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            type_: LongType::S4,
+            endian,
+        }),
+        "s8" => Some(EnumType::Long {
+            type_: LongType::S8,
+            endian,
+        }),
+        _ => return Err(A::Error::custom("invalid type")),
+    };
+    let type_ = type_.ok_or_else(|| A::Error::custom("invalid type"))?;
```
Comment (ArkadiyParovozow) [inline src/kaitai_spec/attribute/enumeration.rs:96]:
```suggestion
    let type_ = AttributeType::Enumeration(Enumeration {
         name,
         type_: match &type_unchecked[0..2] {
        "u1" => Some(EnumType::U1),
        "u2" => Some(EnumType::Long {
            type_: LongType::U2,
            endian,
        }),
        "u4" => Some(EnumType::Long {
            type_: LongType::U4,
            endian,
        }),
        "u8" => Some(EnumType::Long {
            type_: LongType::U8,
            endian,
        }),
        "s1" => Some(EnumType::S1),
        "s2" => Some(EnumType::Long {
            type_: LongType::S2,
            endian,
        }),
        "s4" => Some(EnumType::Long {
            type_: LongType::S4,
            endian,
        }),
        "s8" => Some(EnumType::Long {
            type_: LongType::S8,
            endian,
        }),
        _ => return Err(A::Error::custom("invalid type")),
    }
    });
```

File: src/kaitai_spec/attribute/enumeration.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    let id: String = common_keys
+        .id
+        .ok_or_else(|| A::Error::missing_field("id"))?;
+    return Ok(Attribute {
+        id,
+        doc: common_keys.doc,
+        type_: AttributeType::Enumeration(Enumeration { name, type_ }),
+    });
```
Comment (ArkadiyParovozow) [inline src/kaitai_spec/attribute/enumeration.rs:104]:
```suggestion

    Ok(Attribute {
        id: common_keys
            .id
            .ok_or_else(|| A::Error::missing_field("id"))?,
        doc: common_keys.doc,
        doc_ref: /* ... */,
        type_,
    })
```

**Review by ArkadiyParovozow — CHANGES_REQUESTED** [review]
Please merge origin/main branch too

File: src/kaitai_spec/attribute/enumeration.rs (hunk 1 above)
Comment (MrJonatans) [reply src/kaitai_spec/attribute/enumeration.rs:65]: I’ll remove the extra variable, but is it really necessary to change .chars().count() to .len()?

Comment (ArkadiyParovozow) [reply src/kaitai_spec/attribute/enumeration.rs:65]: Yep, what the concerns do you have?

File: src/kaitai_spec/attribute/enumeration.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    use crate::kaitai_spec::attribute::{
+        enumeration,
+        enumeration::{Endian, EnumType, LongType},
+        Attribute,
+    };
+
+    #[test]
+    fn simple_enum1() {
+        let str: &str = r#"
+        id: hehe
```
[the remark this answers is not in the document]
Comment (MrJonatans) [reply src/kaitai_spec/attribute/enumeration.rs:120]: what does "neutral data" mean? Should I remove the type declaration (: &str)?

[the remark this answers is not in the document]
Comment (ArkadiyParovozow) [reply src/kaitai_spec/attribute/enumeration.rs:120]: I mean using different strings/text for items: `id: item_1`, `id: object2`, etc

File: src/kaitai_spec/attribute/enumeration.rs (hunk 1 above)
Comment (MrJonatans) [reply src/kaitai_spec/attribute/enumeration.rs:65]: if refer to the official documentation:
image

Comment (MrJonatans) [reply src/kaitai_spec/attribute/enumeration.rs:65]: Possibly this could cause conflicts(?)

Comment (ArkadiyParovozow) [reply src/kaitai_spec/attribute/enumeration.rs]: Yep, it is correct, however we check the input against the string in ASCII encoding. So it's ok

**Review by ArkadiyParovozow — APPROVED** [review]

**Pull request merged**